aws/external/resolve_credentials.go

```diff
@@ -0,0 +1,404 @@
+package external
+
+import (
+	"fmt"
+	"net/url"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/awserr"
+	"github.com/aws/aws-sdk-go-v2/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go-v2/aws/ec2rolecreds"
+	"github.com/aws/aws-sdk-go-v2/aws/endpointcreds"
+	"github.com/aws/aws-sdk-go-v2/aws/processcreds"
+	"github.com/aws/aws-sdk-go-v2/aws/stscreds"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+)
+
+const (
+	// valid credential source values
+	credSourceEc2Metadata  = "Ec2InstanceMetadata"
+	credSourceEnvironment  = "Environment"
+	credSourceECSContainer = "EcsContainer"
+)
+
+var (
+	ecsContainerEndpoint = "http://169.254.170.2" // not constant to allow for swapping during unit-testing
+
+	// WebIdentityEmptyRoleARNErr will occur if 'AWS_WEB_IDENTITY_TOKEN_FILE' was set but
+	// 'AWS_ROLE_ARN' was not set.
+	WebIdentityEmptyRoleARNErr = awserr.New(stscreds.ErrCodeWebIdentity, "role ARN is not set", nil)
+
+	// WebIdentityEmptyTokenFilePathErr will occur if 'AWS_ROLE_ARN' was set but
+	// 'AWS_WEB_IDENTITY_TOKEN_FILE' was not set.
+	WebIdentityEmptyTokenFilePathErr = awserr.New(stscreds.ErrCodeWebIdentity, "token file path is not set", nil)
+)
+
+// ResolveCredentials extracts a credential provider from slice of config sources.
+//
+// If an explict credential provider is not found the resolver will fallback to resolving
+// credentials by extracting a credential provider from EnvConfig and SharedConfig.
+func ResolveCredentials(cfg *aws.Config, configs Configs) error {
+	found, err := ResolveCredentialProvider(cfg, configs)
+	if err != nil {
+		return err
+	}
+	if found {
+		return nil
+	}
+
+	err = ResolveCredentialChain(cfg, configs)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ResolveCredentialProvider extracts the first instance of Credentials from the
+// config slices.
+//
+// Config providers used:
+// * CredentialsProviderProvider
+func ResolveCredentialProvider(cfg *aws.Config, configs Configs) (bool, error) {
+	credentials, found, err := GetCredentialsProvider(configs)
+	if err != nil {
+		return false, err
+	}
+	if !found {
+		return false, nil
+	}
+
+	cfg.Credentials = credentials
+
+	return true, nil
+}
+
+// ResolveCredentialChain resolves a credential provider chain using EnvConfig and SharedConfig
+// if present in the slice cof provided configs.
+func ResolveCredentialChain(cfg *aws.Config, configs Configs) (err error) {
+	_, sharedProfileSet, err := GetSharedConfigProfile(configs)
+	if err != nil {
+		return err
+	}
+
+	envConfig, sharedConfig, other := getAWSConfigSources(configs)
+
+	switch {
+	case sharedProfileSet:
+		err = resolveCredsFromProfile(cfg, envConfig, sharedConfig, other)
+	case envConfig.Credentials.HasKeys():
```

Comment:
Nit for readability it would be nice to have a trailing newline after each case block to make little easier to read the case sections.